src/frontend/context/LocationContext.tsx

```diff
@@ -0,0 +1,209 @@
+const STORE_KEY = "@MapeoPosition@1";
+
+export type LocationContextType = {
+  // If available, details of the current position
+  position?: Position;
+  // What location services / providers are available on this device
+  provider?: Provider;
+  // Whether the user has granted permissions to use location to this app
+  permission?: PermissionResult;
+  // This is the previous known position from the last time the app was open
+  savedPosition?: Position | null;
+  // True if there is some kind of error getting the device location
+  error?: boolean;
+};
+
+interface Subscription {
+  remove: () => void;
+}
+
+const DEFAULT_CONTEXT: LocationContextType = {
+  error: false,
+};
+
+const positionOptions = {
+  // See https://docs.expo.io/versions/v32.0.0/sdk/location/#locationaccuracy
+  // this is the best possible accuracy using GPS and other sensors
+  accuracy: Location.Accuracy.BestForNavigation,
+  // This is the interval between location updates. We should get a new GPS
+  // reading every 2000ms.
+  timeInterval: 2000,
+};
+
+// Timeout between location updates --> means location was probably turned off
+// so we need to check it.
+const LOCATION_TIMEOUT = 10000;
+
+const stopWatchingLocation = (
+  timeoutIdRef: React.MutableRefObject<number | undefined>,
+  subscriptionRef: React.MutableRefObject<Subscription | undefined>
+) => {
+  log("Stopping GPS watch");
+  if (subscriptionRef.current) subscriptionRef.current.remove();
+  if (timeoutIdRef.current) window.clearTimeout(timeoutIdRef.current);
+  subscriptionRef.current = undefined;
+};
+
+const LocationContext: React.Context<LocationContextType> = React.createContext<
+  LocationContextType
+>(DEFAULT_CONTEXT);
+
+/**
+ * The LocationProvider provides details about the current device location based
+ * on sensors including GPS. It must be included in the component heirarchy
+ * below the PermissionsProvider, since it needs to read the permissions granted
+ * for device location. There is no event we can listen to for when the user
+ * switches off location (e.g. changes to airplane mode) so we use a timeout ->
+ * if we get not new readings for 10 seconds then we check to see whether the
+ * user has turned off location.
+ */
+const _LocationProvider = ({
+  children,
+  permissions,
+}: React.PropsWithChildren<{
+  permissions: PermissionsType;
+}>) => {
+  const [error, setError] = React.useState(false);
+  const [savedPosition, setSavedPosition] = React.useState<Position>();
+  const [position, setPosition] = React.useState<Position>();
+  const [provider, setProvider] = React.useState<Provider>();
+
+  const timeoutId = React.useRef<number>();
+  const subscription = React.useRef<Subscription>();
+
+  const fineLocationPermissionResult = permissions[
+    PERMISSIONS.ACCESS_FINE_LOCATION
+  ] as PermissionResult;
+
+  const updateStatus = React.useCallback(async () => {
+    try {
+      if (fineLocationPermissionResult !== RESULTS.GRANTED) return;
```

Comment:
technically TS complains about 85-87 and 91 here since `PermissionsContext` isn't converted to TS yet. updating that file is out of the scope of this PR though